engine/crates/federated-graph/src/federated_graph/v2.rs

```diff
@@ -0,0 +1,560 @@
+    pub composed_directives: Directives,
+
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub description: Option<StringId>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone)]
+pub struct InputObject {
+    pub name: StringId,
+
+    pub fields: InputValueDefinitions,
+
+    /// All directives that made it through composition. Notably includes `@tag`.
+    pub composed_directives: Directives,
+
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub description: Option<StringId>,
+}
+
+/// A (start, len) range in FederatedSchema.
+pub type Directives = (DirectiveId, usize);
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct DirectiveId(pub usize);
+
+pub const NO_DIRECTIVES: Directives = (DirectiveId(0), 0);
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct InputValueDefinitionId(pub usize);
+
+/// A (start, len) range in FederatedSchema.
+pub type InputValueDefinitions = (InputValueDefinitionId, usize);
+
+pub const NO_INPUT_VALUE_DEFINITION: InputValueDefinitions = (InputValueDefinitionId(0), 0);
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct EnumValueId(pub usize);
+
+/// A (start, len) range in FederatedSchema.
+pub type EnumValues = (EnumValueId, usize);
+
+pub const NO_ENUM_VALUE: EnumValues = (EnumValueId(0), 0);
+
+impl std::ops::Index<Directives> for FederatedGraphV2 {
+    type Output = [Directive];
+
+    fn index(&self, index: Directives) -> &Self::Output {
+        let (DirectiveId(start), len) = index;
+        &self.directives[start..(start + len)]
+    }
+}
+
+impl std::ops::Index<InputValueDefinitions> for FederatedGraphV2 {
+    type Output = [InputValueDefinition];
+
+    fn index(&self, index: InputValueDefinitions) -> &Self::Output {
+        let (InputValueDefinitionId(start), len) = index;
+        &self.input_value_definitions[start..(start + len)]
+    }
+}
+
+impl std::ops::Index<EnumValues> for FederatedGraphV2 {
+    type Output = [EnumValue];
+
+    fn index(&self, index: EnumValues) -> &Self::Output {
+        let (EnumValueId(start), len) = index;
+        &self.enum_values[start..(start + len)]
+    }
+}
+
+macro_rules! id_newtypes {
+    ($($name:ident + $storage:ident + $out:ident,)*) => {
+        $(
+            impl std::ops::Index<$name> for FederatedGraphV2 {
+                type Output = $out;
+
+                fn index(&self, index: $name) -> &$out {
+                    &self.$storage[index.0]
+                }
+            }
+
+            impl std::ops::IndexMut<$name> for FederatedGraphV2 {
+                fn index_mut(&mut self, index: $name) -> &mut $out {
+                    &mut self.$storage[index.0]
+                }
+            }
+        )*
+    }
+}
+
+id_newtypes! {
+    EnumId + enums + Enum,
+    FieldId + fields + Field,
+    TypeId + field_types + FieldType,
```

Comment:
I think we're reusing `FieldType` from v1 but it shouldn't break in the json format, let's do this.

Reply:
True true, will make it consistent.

Reply:
By the way, probably a change for federated graph v3 but what do you think of making FieldType more like [this](https://github.com/grafbase/grafbase/blob/main/engine/crates/operation-checks/src/schema/wrapping_types.rs)? So instead of interning, use a `DefinitionId` and encode the wrapping types in a `u64`.

Reply:
haha did the same https://github.com/grafbase/grafbase/pull/1324 xD

Reply:
I'm not sure to be honest. We could but we would need a `UnpackedDefinition` for `match` like I did for this: https://github.com/grafbase/grafbase/blob/main/engine/crates/engine-v2/src/response/path.rs#L65

not sure how great that is, but why not 🤷 I'm already reserving the 4 upper bits in all ids in`engine-v2-schema` :D